# Exercise 1 : ANLY 535-51- B-2018/Late Fall - Machine Learning II  

## Smitha Shivakumar
## 11/18/2018


### Problem Statement

Two sensors are installed in Susquehanna river that can measure the level of the water on two different locations. Their measurements goes through a neural network and the output shows if there is a chance of flood or not (y=1 means there is a chance of flood). Suppose sensor 1 shows 0.4 feet while the second one shows 0.32 feet. Is there any chance of flood?

#### Notes for the solution: 
      1. Please use Python to answer to the question
      2. Please generate Random vectors and matrices (using numpy) for weights and bias( seed = 123)
      3. Must explain the final result
    

In [60]:
import numpy as np

# creating the sigmoid function
def activation_function(x,deriv=False):
	if(deriv==True):
	    return x*(1-x)

	return 1/(1+np.exp(-x))



def NeuralNetwork(input_vector):
    
    # running the network with an input vector input_vector.
    input_vector = np.array(input_vector,ndmin=2)
    
    # setting the seed value for the random number geenration
    np.random.seed(123)
    
    #Initializing the weight matrices of the neural network  
    weights_in_hidden = np.random.rand(input_vector.shape[0],4) 
    weights_hidden_out = np.random.rand(4,1) 

    # turning the input vector into a column vector
    input_vector = np.array(input_vector, ndmin=2).T
    
    #calculating the inner hidden layer by multiplying the input vector with the inner hidden weights
    output_vector = np.dot(input_vector, weights_in_hidden)
    output_vector = activation_function(output_vector)

    #calculating the final output yhat by multiplying the hidden layer vector with the outer hidden weights
    output_vector = np.dot(output_vector, weights_hidden_out)
    output_vector = activation_function(output_vector)

    return output_vector

In [61]:
NeuralNetwork([0.4,0.32])

array([[0.82139132],
       [0.81781342]])


### Conculsion:

The output ŷ of a simple 2-layer Neural Network is: ŷ = sigmoid_func(W2 * sigmoid_func(W1*x + b1) + b2)

So the weights W and the biases b are the only variables that affects the output ŷ.

As mentioned in the question 'y=1' means there is a chance of flood

So for our inputs 0.4 and 0.32, our ŷ is 0.8214 and 0.8178 which are less than 1. So, there is a greater than 2/3rd's of a  probability that flood might occur.

Also note there is a marginal higher probability of flood occuring for sensor1 = 0.4 (ŷ = 0.8214) than 0.32 (ŷ = 0.8178)

Running the same Neural Net for a wider array of sensor values below, it is seen that the NeuralNetowrk built in this exercise shows higher probability/output value for higher sensor/input values.


In [62]:
NeuralNetwork([0.4,0.32,0.78,0.93,0.65,0.13])

array([[0.82139132],
       [0.81781342],
       [0.83736548],
       [0.84318336],
       [0.83209654],
       [0.80904069]])